spinalcordtoolbox/reports/qc.py

```diff
@@ -75,9 +75,14 @@ def template(self, mask, ax):
         ax.get_xaxis().set_visible(False)
         ax.get_yaxis().set_visible(False)
 
-    def no_seg_seg(self, mask, ax):
+    def no_seg_seg(self, mask, ax, cmap, norm, fig=None, colorbar=False):
         """Create figure with image overlay. Notably used by sct_registration_to_template"""
-        ax.imshow(mask, cmap='gray', interpolation=self.interpolation, aspect=self.aspect_mask)
+        fig_ax = ax.imshow(mask, cmap=cmap, norm=norm, interpolation=self.interpolation, aspect=self.aspect_mask)
+        if colorbar:
+            cax = ax.inset_axes([1.005, 0.07, 0.011, 0.86])
+            cbar = fig.colorbar(fig_ax, cax=cax, orientation='vertical', pad=0.01, shrink=0.5, aspect=1, ticks=[norm.vmin, norm.vmax])
```

Comment:
Actually, wait, is `fig` even necessary to make a colorbar?

Since we're already passing `fig_ax` and `cax`, we should already be able to identify the correct axes, right? So, would it be possible to just call [`pyplot.colorbar`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.colorbar.html) directly? If this is possible, then that would reduce the number of parameters needed for `no_seg_seg`...

Reply:
Ah! Also, reading through the details of `pyplot.colorbar`, I notice it mentions:

> cax: Axes, optional
> Axes into which the colorbar will be drawn. If [None](https://docs.python.org/3/library/constants.html#None), then a new Axes is created and the space for it will be stolen from the Axes(s) specified in ax.

I'm curious... if the "space for it will be stolen", then does that mean we _don't_ need to manually set dimensions for the axes? (If so, that would simplify things even further!)

Reply:
> Ah! Also, reading through the details of `pyplot.colorbar`, I notice it mentions:
> 
> > cax: Axes, optional
> > Axes into which the colorbar will be drawn. If [None](https://docs.python.org/3/library/constants.html#None), then a new Axes is created and the space for it will be stolen from the Axes(s) specified in ax.
> 
> I'm curious... if the "space for it will be stolen", then does that mean we _don't_ need to manually set dimensions for the axes? (If so, that would simplify things even further!)

This is what I did before, but it made the images reaaly small, so that is why I wanted to specify the size of th axis instead!

Reply:
it works when I call matplolib_pypplot, but I get this warining: 
```
/home/sabeda/spinalcordtoolbox/spinalcordtoolbox/reports/qc.py:83: UserWarning: Adding colorbar to a different Figure <Figure size 500x25 with 1 Axes> than <Figure size 640x480 with 0 Axes> which fig.colorbar is called on.
  cbar = mpl_plt.colorbar(fig_ax, cax=cax, orientation='vertical', pad=0.01, shrink=0.5, aspect=1, ticks=[norm.vmin, norm.vmax])
```

Reply:
done in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4602/commits/7ed632d20bee49cade8dd9b0b8fb161cfea951c5

Reply:
> it works when I call matplolib_pypplot, but I get this warining:

I'm able to reproduce the warning. I've figured out the reason why this warning is thrown, too:

- My suggestion above (creating a colorbar using `pyplot`) requires that our figures be managed by `pyplot`, so that `pyplot` can create a colorbar on the correct figure.
- However, currently `qc.py` is written in the "[explicit (object oriented) style](https://matplotlib.org/stable/users/explain/quick_start.html#the-explicit-and-the-implicit-interfaces)". We create new figures via calls to the class `matplotlib.figure.Figure()`, which _doesn't_ register the figure with `pyplot`.
- To demonstrate further, if we try to manually [set this figure as the current figure](https://stackoverflow.com/a/7987462) with `pyplot`, matplotlib will throw an error: "`ValueError: The passed figure is not managed by pyplot`"
- To fix this, if we call `matplotlib.pyplot.figure()` instead of `figure.Figure()`, the figure will be managed by pyplot and set as the "current figure", and thus `pyplot.colorbar` will be correctly associated with the figure, and no warning will be thrown.

I was initially worried about the risks of mixing "OO-style" matplotlib code with implicit, pyplot-managed code. However, the good news is that even in `matplotlib`'s own example code, they still use the "implicit" `matplotlib.pyplot.figure()` to create the figure.

> So one can use the OO-style
>
> ```python
> # Note that even in the OO-style, we use `.pyplot.figure` to create the Figure.
> fig, ax = plt.subplots(figsize=(5, 2.7), layout='constrained')
> ```

So, I think that means we should use pyplot-managed figures, even if we continue to use the explicit style in most of our code.

((Tagging @mguaypaq as I think "implicit vs explicit" is something we should be aware of when we do our QC refactors. I wasn't really familiar with the nuances here until I did this little deep dive.))

Reply:
Since I updated the code for this PR, I think I'm going to mark this as resolved, but create a new issue to address the remaining examples.